crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2717,6 +2717,21 @@ impl<'db> TypeInferenceBuilder<'db> {
         // - `[ast::CompOp::Is]`: return `false` if unequal, `bool` if equal
         // - `[ast::CompOp::IsNot]`: return `true` if unequal, `bool` if equal
         match (left, right) {
+            (Type::Union(union), other) => Some(UnionType::from_elements(
+                self.db,
+                union
+                    .elements(self.db)
+                    .iter()
+                    .filter_map(|element| self.infer_binary_type_comparison(*element, op, other)),
```

Comment:
I'm not 100% sure it's correct to filter out `None`s here. But it seems wrong to return `None` if there are valid `Some(…)` cases in the overall union of result types.

Reply:
I would vote for falling back to `bool` if one of the elements of the union would lead to the comparison being invalid